### ISSUE TITLE
Misleading error when implementing foreign trait on pointers to local types

### ISSUE THREAD

**Issue by Hoverbear** [rust-lang/rust#99572]
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e0a6148a3b4952ab3abd1f4bd7784029

```rust
use serde::ser::Serialize;

struct BooperDoop;

impl Serialize for *mut BooperDoop {
    // ...
}

fn main() {}
```

The current output is:

```rust
Compiling playground v0.0.1 (/playground)
error[[E0117]](https://doc.rust-lang.org/stable/error-index.html#E0117): only traits defined in the current crate can be implemented for arbitrary types
 --> src/main.rs:5:1
  |
5 | impl Serialize for *mut BooperDoop {
  | ^^^^^^^^^^^^^^^^^^^---------------
  | |                  |
  | |                  `*mut BooperDoop` is not defined in the current crate
  | impl doesn't use only types from inside the current crate
  |
  = note: define and implement a trait or new type instead

For more information about this error, try `rustc --explain E0117`.
error: could not compile `playground` due to previous error
```

Ideally the output should look like:

```rust
Compiling playground v0.0.1 (/playground)
error[[E0117]](https://doc.rust-lang.org/stable/error-index.html#E0117): only traits defined in the current crate can be implemented for arbitrary types
 --> src/main.rs:5:1
  |
5 | impl Serialize for *mut BooperDoop {
  | ^^^^^^^^^^^^^^^^^^^---------------
  | |                  |
  | |                  `*mut BooperDoop` is not defined in the current crate, try implementing on `Booperdoop`
  | impl doesn't use only types from inside the current crate
  |
  = note: define and implement a trait or new type instead

For more information about this error, try `rustc --explain E0117`.
error: could not compile `playground` due to previous error
```

**Comment by compiler-errors** [comment]
> impl doesn't use only types from inside the current crate

This could use some work I think. Maybe "impl mentions type from outside the current crate" could work, though we should probably make sure that wording is compatible with current coherence rules.

Though, in this case, for a non-`#[fundamental]` self type like `*mut _`, we could then replace that^ message with something even more specific.

I'm a bit skeptical of mentioning "try implementing `Trait` on `YourLocalType`" when we have `impl Trait for *mut YourLocalType`, since implementing something for `T` and `*mut T` can have very different meanings depending on the signatures inside the trait.

**Comment by vincenzopalazzo** [comment]
Let's improve this case!

@rustbot claim

**Comment by Hoverbear** [comment]
This was actually closer to my original error I got, it uses a type alias to create some extra confusion:

```rust
use serde::ser::Serialize;

struct BooperDoop;
type ScooperSwoop = *mut BooperDoop;

impl Serialize for ScooperSwoop {
    // ...
}

fn main() {}
```

**Comment by vincenzopalazzo** [comment]
> This was actually closer to my original error I got, it uses a type alias to create some extra confusion:

Amazing example :) thanks, I'm on it

**Comment by vincenzopalazzo** [comment]
This is a reproducible example without serde!

```rust
use std::fmt;

struct BooperDoop;

impl fmt::Display for *mut BooperDoop {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "This not compile")
    }
}

fn main() {}
```
